File: scripts/local-multi-hypervisor-kvm-smoke.py
#!/usr/bin/env python3
"""Run a bounded KVM-backed local multi-hypervisor replay-readiness smoke rail."""

from __future__ import annotations

import argparse
import json
import os
import shlex
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path

DEFAULT_WORKLOADS = ("raft", "rust-workload")


def run(command: list[str]) -> subprocess.CompletedProcess[str]:
    return subprocess.run(command, text=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)


def require_kvm() -> None:
    kvm = Path("/dev/kvm")
    if not kvm.exists():
        raise SystemExit("local multi-hypervisor KVM smoke requires /dev/kvm")
    if not os.access(kvm, os.R_OK | os.W_OK):
        raise SystemExit("local multi-hypervisor KVM smoke requires read/write access to /dev/kvm")


def parse_workloads(raw: str) -> list[str]:
    workloads = [item.strip() for item in raw.split(",") if item.strip()]
    if len(workloads) < 2:
        raise SystemExit("--workloads must select at least two workloads for multi-hypervisor smoke")
    if len(set(workloads)) != len(workloads):
        raise SystemExit("--workloads must not contain duplicate workloads")
    return workloads


def build_plan(
    *,
    out: Path,
    replay_readiness: str,
    workloads: list[str],
    dogfood_extra: list[str],
) -> dict[str, object]:
    entries = []
    for idx, workload in enumerate(workloads, start=1):
        receipt = out / "run-receipts" / f"{idx:02d}-{workload}-replay-readiness.json"
        dogfood_out = out / "dogfood" / f"{idx:02d}-{workload}"
        command = [
            replay_readiness,
            "--receipt",
            str(receipt),
            "--dogfood",
            workload,
            "--",
            "--output",
            str(dogfood_out),
            *dogfood_extra,
        ]
        entries.append(
            {
                "queue_entry_id": f"kvm-mhq-{idx:04d}",
                "run_id": f"kvm-mh-run-{idx:04d}",
                "workload": workload,
                "command": " ".join(shlex.quote(part) for part in command),
                "receipt_path": str(receipt),
            }
        )
    return {
        "schema_version": 1,
        "campaign_id": "local-kvm-smoke-0001",
        "max_hypervisors": min(len(workloads), 2),
        "state_path": str(out / "campaign-state.json"),
        "artifact_index_path": str(out / "artifact-index.json"),
        "follow_up_policy": {"enabled": False, "reproduce": False, "minimize": False},
        "hypervisors": [
            {
                "hypervisor_worker_id": f"local-kvm-hv-{idx}",
                "node_id": f"local-kvm-node-{idx}",
                "resource_budget": {"vcpus": 2, "memory_mib": 1024},
                "artifact_root": str(out / "hypervisors" / f"local-kvm-hv-{idx}"),
            }
            for idx in range(1, len(workloads) + 1)
        ],
        "queue": {"entries": entries},
        "operator_decisions": [str(out / "operator-decision-receipt.json")],
    }


def write_summary(out: Path, summary: str, receipt_path: Path, plan_path: Path, rc: int, command_output: str) -> None:
    summary_path = out / "summary.txt"
    summary_path.write_text(
        "\n".join(
            [
                "local multi-hypervisor KVM smoke",
                f"status={'passed' if rc == 0 else 'failed'}",
                f"summary={summary.strip() if summary.strip() else '<none>'}",
                f"plan={plan_path}",
                f"receipt={receipt_path}",
                f"queue_state={out / 'campaign-state.json'}",
                "scope=bounded-local-kvm-multi-hypervisor-not-hosted-not-shared-remote-queue-not-cross-machine",
                "raw_log_scraping=false",
                "",
                "runner output:",
                command_output.strip(),
                "",
            ]
        )
    )


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--out", type=Path, default=Path("dogfood-results/local-multi-hypervisor-kvm-smoke-latest"))
    parser.add_argument("--workloads", default=",".join(DEFAULT_WORKLOADS))
    parser.add_argument("--replay-readiness", default=os.environ.get("REPLAY_READINESS", "replay-readiness"))
    parser.add_argument(
        "--scheduler-receipt",
        default=os.environ.get("REPLAY_READINESS_SCHEDULER_RECEIPT", "replay-readiness-scheduler-receipt"),
    )
    parser.add_argument("dogfood_extra", nargs=argparse.REMAINDER, help="extra args passed after each dogfood --output")
    args = parser.parse_args()

    require_kvm()
    out = args.out.resolve()
    if out.exists() and any(out.iterdir()):
        raise SystemExit(f"output directory is not empty: {out}")
    out.mkdir(parents=True, exist_ok=True)
    (out / "run-receipts").mkdir()
    (out / "dogfood").mkdir()

    dogfood_extra = args.dogfood_extra
    if dogfood_extra and dogfood_extra[0] == "--":
        dogfood_extra = dogfood_extra[1:]

    workloads = parse_workloads(args.workloads)
    plan = build_plan(out=out, replay_readiness=args.replay_readiness, workloads=workloads, dogfood_extra=dogfood_extra)
    plan_path = out / "campaign-plan.json"
    receipt_path = out / "campaign-receipt.json"
    plan_path.write_text(json.dumps(plan, indent=2, sort_keys=True) + "\n")

    started = datetime.now(timezone.utc).isoformat()
    command = [args.scheduler_receipt, "--run-multi-hypervisor-plan", str(plan_path), "--output", str(receipt_path)]
    completed = run(command)
    summary = completed.stdout.strip().splitlines()[-1] if completed.stdout.strip() else ""
    write_summary(out, summary, receipt_path, plan_path, completed.returncode, completed.stdout)
    (out / "metadata.json").write_text(
        json.dumps(
            {
                "schema_version": 1,
                "command": "local-multi-hypervisor-kvm-smoke",
                "started_at": started,
                "finished_at": datetime.now(timezone.utc).isoformat(),
                "exit_code": completed.returncode,
                "workloads": workloads,
                "artifacts": {
                    "plan": str(plan_path),
                    "receipt": str(receipt_path),
                    "queue_state": str(out / "campaign-state.json"),
                    "summary": str(out / "summary.txt"),
                },
                "scope": "bounded local KVM multi-hypervisor campaign smoke only; not a hosted service, not a shared remote queue, not cross-machine scheduling, not fleet-scale throughput",
            },
            indent=2,
            sort_keys=True,
        )
        + "\n"
    )
    sys.stdout.write(completed.stdout)
    if completed.returncode == 0:
        print(f"local multi-hypervisor KVM smoke artifacts: {out}")
    else:
        print(f"local multi-hypervisor KVM smoke failed; artifacts: {out}", file=sys.stderr)
    return completed.returncode


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/chaoscontrol-evidence/src/bin/local-multi-hypervisor-kvm-smoke.rs
use std::env;
use std::fs;
use std::os::unix::fs::OpenOptionsExt;
use std::path::{Path, PathBuf};
use std::process::ExitCode;
use std::time::{SystemTime, UNIX_EPOCH};

use chaoscontrol_evidence::rust_automation::local_kvm::{campaign_plan, parse_workloads};
use chaoscontrol_evidence::rust_automation::time::rfc3339_utc;
use chaoscontrol_evidence::typed_operator_command::{
    CommandPlan, EnvironmentMode, EnvironmentSpec, LimitSpec, StdinSpec, TerminationScope,
    MECHANISM_REVISION, PLAN_SCHEMA,
};
use chaoscontrol_evidence::{execute_typed_operator_command_captured, observe_typed_executable};
use serde_json::{json, Value};

const DEFAULT_OUT: &str = "dogfood-results/local-multi-hypervisor-kvm-smoke-latest";
const DEFAULT_WORKLOADS: &str = "raft,rust-workload";
const DEFAULT_REPLAY: &str = "replay-readiness";
const DEFAULT_SCHEDULER: &str = "replay-readiness-scheduler-receipt";
const EXECUTABLE_MAX_BYTES: u64 = 1_073_741_824;
const COMMAND_TIMEOUT_MS: u64 = 30_000;
const ORCHESTRATOR_TIMEOUT_MS: u64 = 900_000;
const STDIN_MAX_BYTES: u64 = 1_024;
const OUTPUT_MAX_BYTES: u64 = 8_388_608;
const POLL_INTERVAL_MS: u64 = 10;
const TEARDOWN_TIMEOUT_MS: u64 = 1_000;
const USAGE_EXIT: u8 = 2;

#[derive(Debug)]
struct Options {
    out: PathBuf,
    workloads: Vec<String>,
    replay_readiness: PathBuf,
    scheduler: PathBuf,
    dogfood_extra: Vec<String>,
}

fn main() -> ExitCode {
    match run(env::args().skip(1).collect()) {
        Ok(code) => ExitCode::from(u8::try_from(code).unwrap_or(1)),
        Err(error) => {
            eprintln!("{error}");
            ExitCode::from(USAGE_EXIT)
        }
    }
}

fn run(args: Vec<String>) -> Result<i32, String> {
    require_kvm()?;
    let options = parse_args(&args)?;
    let out = absolute(&options.out)?;
    prepare_output(&out)?;
    let started_at = now()?;
    let replay_executable =
        observe_typed_executable(&options.replay_readiness, EXECUTABLE_MAX_BYTES)
            .map_err(|error| error.to_string())?;
    let scheduler_executable = observe_typed_executable(&options.scheduler, EXECUTABLE_MAX_BYTES)
        .map_err(|error| error.to_string())?;
    let command_plans = options
        .workloads
        .iter()
        .enumerate()
        .map(|(index, workload)| {
            let number = index + 1;
            let receipt = out
                .join("run-receipts")
                .join(format!("{number:02}-{workload}-replay-readiness.json"));
            let mut args = vec![
                String::from("--receipt"),
                receipt.display().to_string(),
                String::from("--dogfood"),
                workload.clone(),
                String::from("--"),
                String::from("--output"),
                out.join("dogfood")
                    .join(format!("{number:02}-{workload}"))
                    .display()
                    .to_string(),
            ];
            args.extend(options.dogfood_extra.clone());
            let plan = command_plan(replay_executable.clone(), args, COMMAND_TIMEOUT_MS, vec![0]);
            let value = serde_json::to_value(&plan).map_err(|error| error.to_string())?;
            write_json(&out.join(format!("command-plan-{number:02}.json")), &value)?;
            Ok(value)
        })
        .collect::<Result<Vec<_>, String>>()?;
    let plan_path = out.join("campaign-plan.json");
    let receipt_path = out.join("campaign-receipt.json");
    let plan = campaign_plan(&out, &options.workloads, &command_plans)?;
    write_json(&plan_path, &plan)?;

    let scheduler_plan = command_plan(
        scheduler_executable,
        vec![
            String::from("--run-multi-hypervisor-plan"),
            plan_path.display().to_string(),
            String::from("--output"),
            receipt_path.display().to_string(),
        ],
        ORCHESTRATOR_TIMEOUT_MS,
        vec![0, 1, 2],
    );
    let execution_root =
        env::current_dir().map_err(|error| format!("current directory: {error}"))?;
    let captured = execute_typed_operator_command_captured(&scheduler_plan, &execution_root)
        .map_err(|error| error.to_string())?;
    let exit_code = captured.observation.exit_code.unwrap_or(1);
    let stdout = String::from_utf8_lossy(&captured.stdout);
    let stderr = String::from_utf8_lossy(&captured.stderr);
    let command_output = format!("{stdout}{stderr}");
    let summary = command_output.lines().last().unwrap_or("");
    write_summary(
        &out,
        summary,
        &receipt_path,
        &plan_path,
        exit_code,
        &command_output,
    )?;
    write_json(
        &out.join("metadata.json"),
        &json!({
            "schema_version": 1,
            "command": "local-multi-hypervisor-kvm-smoke",
            "started_at": started_at,
            "finished_at": now()?,
            "exit_code": exit_code,
            "workloads": options.workloads,
            "artifacts": {
                "plan": plan_path,
                "receipt": receipt_path,
                "queue_state": out.join("campaign-state.json"),
                "summary": out.join("summary.txt"),
            },
            "scope": "bounded local KVM multi-hypervisor campaign smoke only; not a hosted service, not a shared remote queue, not cross-machine scheduling, not fleet-scale throughput",
        }),
    )?;
    print!("{command_output}");
    if exit_code == 0 {
        println!(
            "local multi-hypervisor KVM smoke artifacts: {}",
            out.display()
        );
    } else {
        eprintln!(
            "local multi-hypervisor KVM smoke failed; artifacts: {}",
            out.display()
        );
    }
    Ok(exit_code)
}

fn command_plan(
    executable: chaoscontrol_evidence::typed_operator_command::ExecutableRef,
    args: Vec<String>,
    timeout_ms: u64,
    accepted_exit_codes: Vec<i32>,
) -> CommandPlan {
    CommandPlan {
        schema: String::from(PLAN_SCHEMA),
        mechanism_revision: String::from(MECHANISM_REVISION),
        executable,
        args,
        working_directory: String::from("."),
        environment: EnvironmentSpec {
            mode: EnvironmentMode::Clear,
            entries: Vec::new(),
        },
        stdin: StdinSpec::Null,
        limits: LimitSpec {
            timeout_ms,
            stdin_max_bytes: STDIN_MAX_BYTES,
            stdout_max_bytes: OUTPUT_MAX_BYTES,
            stderr_max_bytes: OUTPUT_MAX_BYTES,
            poll_interval_ms: POLL_INTERVAL_MS,
            teardown_timeout_ms: TEARDOWN_TIMEOUT_MS,
        },
        accepted_exit_codes,
        reject_stdout_truncation: true,
        reject_stderr_truncation: true,
        termination_scope: TerminationScope::ProcessGroup,
        evidence_eligible: true,
    }
}

fn parse_args(args: &[String]) -> Result<Options, String> {
    let mut out = PathBuf::from(DEFAULT_OUT);
    let mut workloads = String::from(DEFAULT_WORKLOADS);
    let mut replay = env::var("REPLAY_READINESS").unwrap_or_else(|_| String::from(DEFAULT_REPLAY));
    let mut scheduler = env::var("REPLAY_READINESS_SCHEDULER_RECEIPT")
        .unwrap_or_else(|_| String::from(DEFAULT_SCHEDULER));
    let mut dogfood_extra = Vec::new();
    let mut index = 0;
    while index < args.len() {
        match args[index].as_str() {
            "--out" | "--workloads" | "--replay-readiness" | "--scheduler-receipt" => {
                let value = args
                    .get(index + 1)
                    .ok_or_else(|| format!("{} requires a value", args[index]))?;
                match args[index].as_str() {
                    "--out" => out = PathBuf::from(value),
                    "--workloads" => workloads = value.clone(),
                    "--replay-readiness" => replay = value.clone(),
                    _ => scheduler = value.clone(),
                }
                index += 2;
            }
            "--" => {
                dogfood_extra.extend_from_slice(&args[index + 1..]);
                break;
            }
            value => {
                dogfood_extra.push(value.to_string());
                index += 1;
            }
        }
    }
    let workloads = parse_workloads(&workloads)?;
    Ok(Options {
        out,
        workloads,
        replay_readiness: PathBuf::from(replay),
        scheduler: PathBuf::from(scheduler),
        dogfood_extra,
    })
}

fn require_kvm() -> Result<(), String> {
    fs::OpenOptions::new()
        .read(true)
        .write(true)
        .custom_flags(libc::O_CLOEXEC)
        .open("/dev/kvm")
        .map(|_| ())
        .map_err(|_| {
            String::from("local multi-hypervisor KVM smoke requires read/write access to /dev/kvm")
        })
}

fn absolute(path: &Path) -> Result<PathBuf, String> {
    if path.is_absolute() {
        Ok(path.to_path_buf())
    } else {
        Ok(env::current_dir()
            .map_err(|error| format!("current directory: {error}"))?
            .join(path))
    }
}

fn prepare_output(out: &Path) -> Result<(), String> {
    if out.exists()
        && fs::read_dir(out)
            .map_err(|error| format!("{}: {error}", out.display()))?
            .next()
            .is_some()
    {
        return Err(format!("output directory is not empty: {}", out.display()));
    }
    fs::create_dir_all(out).map_err(|error| format!("{}: {error}", out.display()))?;
    fs::create_dir(out.join("run-receipts"))
        .map_err(|error| format!("{}: {error}", out.display()))?;
    fs::create_dir(out.join("dogfood")).map_err(|error| format!("{}: {error}", out.display()))
}

fn write_summary(
    out: &Path,
    summary: &str,
    receipt: &Path,
    plan: &Path,
    exit_code: i32,
    command_output: &str,
) -> Result<(), String> {
    let status = if exit_code == 0 { "passed" } else { "failed" };
    let summary = if summary.trim().is_empty() {
        "<none>"
    } else {
        summary.trim()
    };
    let text = format!(
        "local multi-hypervisor KVM smoke\nstatus={status}\nsummary={summary}\nplan={}\nreceipt={}\nqueue_state={}\nscope=bounded-local-kvm-multi-hypervisor-not-hosted-not-shared-remote-queue-not-cross-machine\nraw_log_scraping=false\n\nrunner output:\n{}\n",
        plan.display(), receipt.display(), out.join("campaign-state.json").display(), command_output.trim()
    );
    fs::write(out.join("summary.txt"), text)
        .map_err(|error| format!("{}: {error}", out.join("summary.txt").display()))
}

fn write_json(path: &Path, value: &Value) -> Result<(), String> {
    let mut bytes = serde_json::to_vec_pretty(value)
        .map_err(|error| format!("{}: encode failed: {error}", path.display()))?;
    bytes.push(b'\n');
    fs::write(path, bytes).map_err(|error| format!("{}: {error}", path.display()))
}

fn now() -> Result<String, String> {
    let seconds = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|error| format!("system clock before Unix epoch: {error}"))?
        .as_secs();
    rfc3339_utc(seconds)
}
